test/algorithms/test_qft.cpp

```diff
@@ -69,7 +59,13 @@ TEST_P(QFT, Functionality) {
   // there should be no error constructing the circuit
   ASSERT_NO_THROW({ qc = std::make_unique<qc::QFT>(nqubits, false); });
   // there should be no error building the functionality
-  ASSERT_NO_THROW({ func = buildFunctionality(qc.get(), dd); });
+  auto exp = std::make_unique<dd::FunctionalityConstructionExperiment>();
+  exp->dd = std::make_unique<dd::Package<>>(nqubits);
+  initialCacheCount = exp->dd->cn.cacheCount();
+  initialComplexCount = exp->dd->cn.realCount();
```

Comment:
No real need to construct a duplicate package instance here.
The initial Cache and Complex Count should be compile time constants.
If I am not mistaken, the initial complex count should be `1` (the 0.5 entry) and the initial cache count should be `0`.
So, technically, you can fully remove these `initial...` member from this file (and also the Grover file, where similar comments apply).

Reply:
I removed both of the initial members from both of the classes.
It wasn't possible to make exp a member in these classes. A smart pointer of SimulationExperiment (or FunctionalityConstructionExperiment) and a smart pointer of Experiment weren't compatible. exp->sim or ->func gave errors because `Experiment`s aren't supposed to have those.

Reply:
Such pointers _are_ compatible, it just requires runtime casting using `dynamic_cast` to the appropriate class.
If you have a base class pointer `std::unique_ptr<Base> p` and a derived class `Derived`, and you know that `p` is pointing to a derived class instance, you can do
```c++
auto* derived = dynamic_cast<Derived*>(p.get());
assert(derived != nullptr); // to make sure the cast succeeded
```